src/main/java/org/candlepin/auth/CloudRegistrationAuth.java

```diff
@@ -267,23 +275,142 @@ public String generateRegistrationToken(Principal principal, CloudRegistrationIn
+     *     if the cloud registration details are null, incomplete, or malformed
+     *
+     * @return cloud authentication results including the JWT token, owner ID, and or the anonymous
+     * consumer UUID
+     */
+    public CloudAuthenticationResultDTO generateRegistrationTokenV2(Principal principal,
+        CloudRegistrationInfo cloudRegInfo)
+        throws CloudRegistrationAuthorizationException, MalformedCloudRegistrationException {
+
+        if (!this.enabled) {
+            throw new UnsupportedOperationException(
+                "cloud registration is not enabled on this Candlepin instance");
+        }
+
+        if (principal == null) {
+            throw new IllegalArgumentException("principal is null");
+        }
+
+        if (cloudRegInfo == null) {
+            throw new IllegalArgumentException("cloudRegistrationInfo is null");
+        }
+
+        CloudAuthenticationResult result = this.cloudRegistrationAdapter
+            .resolveCloudRegistrationDataV2(cloudRegInfo);
+
+        String cloudAccountId = result.getCloudAccountId();
+        if (cloudAccountId == null || cloudAccountId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider account could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String cloudInstanceId = result.getCloudInstanceId();
+        if (cloudInstanceId == null || cloudInstanceId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud instance ID could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String cloudProviderShortName = result.getCloudProviderShortName();
+        if (cloudProviderShortName == null || cloudProviderShortName.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("cloud provider short name could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String productId = result.getProductId();
+        if (productId == null || productId.isBlank()) {
+            String errmsg = this.i18nProvider.get()
+                .tr("product id could not be resolved");
+
+            throw new CloudRegistrationAuthorizationException(errmsg);
+        }
+
+        String ownerKey = result.getOwnerKey();
+        boolean ownerAndPoolExists = false;
+        if (ownerKey != null && !ownerKey.isBlank() && result.isEntitled()) {
+            Owner owner = ownerCurator.getByKey(ownerKey);
+            if (owner != null) {
+                ownerAndPoolExists = !poolCurator.listByOwnerAndProduct(owner, productId).isEmpty();
+            }
+        }
+
+        String anonymousConsumerUuid = null;
+        if (ownerKey == null || ownerKey.isBlank() || !result.isEntitled() || !ownerAndPoolExists) {
+
+            // TODO: Schedule anonymous org creation job
```

Comment:
Unfortunately I don't see a way of verifying that the job was scheduled through the spec tests. Ideally we would be able to use the GET `/jobs` endpoint and specify the owner key that is unique to the spec test and the job key. However, looking at cp_async_jobs table we don't actually store the owner id for these jobs in the `owner_id` column.

So, without being able to filter the jobs down by owner, I don't think it's possible to reliably tell if a job was scheduled by that particular test because there will be concurrently running spec tests also creating jobs for the job key `CloudAccountOrgSetupJob`. However, I did add assertions for scheduling the job in `CloudRegistrationResourceTest.java`.

Reply:
As I mentioned in one of my other comments, when queueing the job, we should be also setting the owner key (if it is not null, when sent by the adapters). so at least in 1 use case (when we're running this job and the org exists, but isn't entitled upstream yet), we could check for the specific job being completed